### PR TITLE
Improve lifetime name annotations for closures & async functions

### PR DESCRIPTION
**Pull request by SNCPlay42**
* Don't refer to async functions as "generators" in error output
* Where possible, emit annotations pointing exactly at the `&` in the return type of closures (when they have explicit return types) and async functions, like we do for arguments.
Addresses #74072, but I wouldn't call that *closed* until annotations are identical for async and non-async functions.
* Emit a better annotation when the lifetime doesn't appear in the full name type, which currently happens for opaque types like `impl Future`. Addresses #74497, but further improves could probably be made (why *doesn't* it appear in the type as `impl Future + '1`?)
This is included in the same PR because the changes to `give_name_if_anonymous_region_appears_in_output` would introduce ICE otherwise (it would return `None` in cases where it didn't previously, which then gets `unwrap`ped)

### CODE REVIEW

File: compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs [hunk 1]
```diff
@@ -637,50 +675,131 @@ impl<'tcx> MirBorrowckCtxt<'_, 'tcx> {
...
+    ///
+    /// ```
+    /// async fn foo() -> i32 {}
+    /// ```
+    ///
+    /// this function, given the lowered return type of `foo`, an [`OpaqueDef`] that implements `Future<Output=i32>`,
+    /// returns the `i32`.
+    ///
+    /// [`OpaqueDef`]: hir::TyKind::OpaqueDef
+    fn get_future_inner_return_ty(&self, hir_ty: &'tcx hir::Ty<'tcx>) -> &'tcx hir::Ty<'tcx> {
```
Comment (SNCPlay42) [inline compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs:762]: Should this live somewhere more general, like on `rustc_middle::hir::map::Map`?

**Comment by Dylan-DPC** [comment]
@SNCPlay42 if you can resolve the new conflicts, we can get it reviewed after that to proceed with this

**Comment by Dylan-DPC** [comment]
@estebank this is ready for review

**Comment by tmandry** [comment]
Just to check, does this fix #75785?

**Comment by SNCPlay42** [comment]
@tmandry no, this pr would change the output to what is shown in https://github.com/rust-lang/rust/issues/75785#issuecomment-678480523 which still points at the wrong reference.

File: compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs [hunk 2]
```diff
@@ -6,8 +6,10 @@ use rustc_hir::def::{DefKind, Res};
 use rustc_middle::ty::print::RegionHighlightMode;
 use rustc_middle::ty::subst::{GenericArgKind, SubstsRef};
 use rustc_middle::ty::{self, RegionVid, Ty};
-use rustc_span::symbol::kw;
-use rustc_span::{symbol::Symbol, Span, DUMMY_SP};
+use rustc_span::{
+    symbol::{kw, sym, Ident, Symbol},
+    Span, DUMMY_SP,
+};
```
Comment (Mark-Simulacrum) [inline compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs:12]: Please don't split use statements to multiple lines, instead expand this to two use statements (as before).

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with nit fixed

File: compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs (hunk 1 above)
Comment (guswynn) [reply compiler/rustc_mir/src/borrow_check/diagnostics/region_name.rs:766]: #78755
This one is actually very close to the one in #76765, basically only differing in how we obtain the OpaqueDef that represents the Future that is the return type of the async fn

**Pull request merged**